server/cmwell-data-tools/src/main/scala/cmwell/tools/data/utils/akka/Retry.scala

```diff
@@ -186,18 +188,28 @@ object Retry extends DataToolsLogging with DataToolsConfig{
           e.discardBytes()
           Future.successful(Failure(new Exception(s"status is not success ($s) $e")) -> state.copy(response = response.toOption))
 
-        case (response@Success(res@HttpResponse(s, _, e, _)), state) =>
+        case (response@Success(res@HttpResponse(s, headers, e, _)), state) =>
           // consume HTTP response bytes and later pack them in fake response
-          val responseAndState = e.withoutSizeLimit().dataBytes.runFold(blank)(_ ++ _)
-            .map { dataBytes =>
-              val isValidBody = responseBodyValidator(dataBytes)
-
-              if (isValidBody) {
-                Success(res.copy(entity = dataBytes)) -> state.copy(response = response.toOption)
-              } else {
-                Failure(new Exception("response body is not valid")) -> state.copy(response = response.toOption)
+          val responseAndState = (e match {
+            case chunkedEntity:Chunked => {
+              chunkedEntity.chunks.runFold[ (ByteString,Option[Seq[HttpHeader]]) ]((ByteString.empty,None)) {
+                case (accumulatedEntity, chunk: LastChunk) => (accumulatedEntity._1 ++ chunk.data, Option(chunk.trailer))
+                case (accumulatedEntity, chunk) => (accumulatedEntity._1 ++ chunk.data, None)
               }
             }
+            case entity:HttpEntity => {
+              entity.withoutSizeLimit.dataBytes.runFold(blank)(_ ++ _).map (_ -> None)
+            }
+          }).map {
+            case (entityBytes, trailerHeaders) =>
+
+              val combinedHeaders = headers ++ trailerHeaders.getOrElse(Seq.empty)
+
+              responseValidator(entityBytes, combinedHeaders) match {
```

Comment:
response validator hides the true reason why it failed.
maybe the function should get a response and return Try with Success(response) or Failure(failure reason)

It doesn't matter much how it will be implemented just I want the logs to be clear with the exact failure reason (unless it's written some other place)

Reply:
What about function returns Try[Boolean], or maybe Either[Boolean, Seq[String]] (sequence of String being failure reasons...?

Reply:
Gilad likes Either, I less.
A Seq of failures is a good idea but I want somehow to expose it to the "outer" world (e.g. logs and such). Usually you have the ability to propagate an exception.
If you find a way to expose all the error it's better.
Try[Boolean] is the simplest but I prefer Try[Unit]. You won't use the value anyway.